### PR TITLE
ping cluster before make plugin ready, refactor some code

### PR DESCRIPTION
**Pull request by StarpTech**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -3,22 +3,37 @@
 const fp = require('fastify-plugin')
 const elasticsearch = require('elasticsearch')
 
-function createClient (options) {
```
Comment (StarpTech) [inline index.js]: I see no need to wrap it in a function we still have `fastifyElasticSearch` as stack trace.

File: index.js [hunk 2]
```diff
@@ -3,22 +3,37 @@
...
+
+  client.ping(
+    {
+      // ping usually has a 3000ms timeout
+      requestTimeout
+    },
+    function (err) {
+      if (err) {
+        fastify.log.error(err, 'elasticsearch cluster is down!')
+        throw err
```
Comment (mcollina) [inline index.js]: do not log, call `next(err)`.

File: index.js [hunk 3]
```diff
@@ -3,22 +3,37 @@
...
     .addHook('onClose', closeESClient)
 
-  next()
-}
+  function closeESClient () {
+    fastify.log.debug('elasticsearch client is closing ...')
+    return client.close()
+  }
 
-function closeESClient (fastify) {
```
Comment (StarpTech) [inline index.js]: I think its cleaner to stay in the closure rather access the instance over decorators.

File: index.js (hunk 2 above)
Comment (StarpTech) [reply index.js]: oh, forgot it.

File: index.js (hunk 3 above)
Comment (mcollina) [reply index.js]: yeah it's ok.

**Comment by StarpTech** [comment]
Done!

File: test.js [hunk 4]
```diff
@@ -36,5 +36,21 @@ t.test('fastify-elasticsearch', t => {
...
+  t.test('with unreachable cluster', t => {
+    t.plan(1)
+
+    const client = new elasticsearch.Client({ host: '127.0.0.1:9999' })
+
+    const fastify = Fastify()
+    fastify.register(fastifyElasticSearch, { client })
+
+    fastify.ready()
+      .then(e => t.fail(e))
```
Comment (mcollina) [inline test.js]: Then will have no argument, and definitely not an exception here. Should be something like `t.fail('should not boot successfully')`.

Comment (StarpTech) [reply test.js]: Good catch

**Comment by StarpTech** [comment]
Should be fine now

File: index.js [hunk 5]
```diff
@@ -3,22 +3,34 @@
...
     .addHook('onClose', closeESClient)
 
-  next()
-}
-
-function closeESClient (fastify) {
-  return fastify.elasticsearch.close()
+  function closeESClient () {
+    fastify.log.debug('elasticsearch client is closing ...')
+    return client.close()
```
Comment (delvedor) [inline index.js]: Is this a sync function?
If not you need to handle the `done` callback.
https://github.com/mcollina/avvio#apponclosefunccontext-done

Comment (StarpTech) [reply index.js]: It is sync https://github.com/elastic/elasticsearch-js/blob/14.x/src/lib/client.js#L54

**Pull request merged**